version docs (#451)
elastic/docs-builder: docs/versions/index.md

---
navigation_title: "Versions and types"
---

# Documenting versions and deployment types

In the new documentation system, we document features in a centralized place, regardless of the software version or deployment type it applies to. 
For example, we might document the serverless and v9 implementations of a feature on a single page, and then use content patterns to highlight where prerequisites, options, or behaviors differ between these implementations.

This approach improves our documentation in several ways: 

* When a user arrives in our documentation from an outside source, they'll be likelier to land on a page or section that applies to them.
* There is a single "source of truth" for each feature, which helps us to maintain consistency, accuracy, and maintainability of our documentation over time, and avoids "drift" between multiple similar sets of documentation.
* Comparing and contrasting differences helps users to understand what is available to them, and improves awareness of the ways certain offerings might improve their experience.

> ::::{note}

Reviewer: I think it will definitely need to be a living document, especially as the outstanding enhancement requests are resolved and new version releases (9.1) or versioning schemes (e.g. for serverless) need to be considered. I think since this is a public repo and ideally visible to all our internal and external contributors, we should ultimately set a codeowner so that changes are always reviewed by someone from docs. If that's not what you're asking and there's some other place ownership needs to be tracked, lmk!

This page documents the first version of our approach to documenting differences in versions and deployment types, using our current technologies. 
Our approach might change as additional documentation features become available.
::::

## Basic concepts and principles

* [Versioning facets](#versioning-facets)
* [Defaults and hierarchy](#defaults-and-hierarchy)
* [Versions and lifecycle states](#versions-and-lifecycle-states)

### Versioning facets
There are multiple facets to versioning that we need to consider: 

| Facet | Description |
| --- | --- |

> | **Stack flavor** | The Elasticsearch or Kibana feature base used for basic functionality. Either **Serverless** (sometimes "Elastic Cloud Serverless") or **{{stack}} <version>** |

Author: See https://github.com/elastic/docs-builder/pull/451/files/7f71ed6cb11d9cb537a7d065af44d74f50087ff0#r1951784255

> | **Stack flavor** | The Elasticsearch or Kibana feature base used for basic functionality. Either **Serverless** (sometimes "Elastic Cloud Serverless") or **{{stack-short}} <version>** |

Reviewer 2: This isn't really a Stack flavor. It specifies the  Elastic Stack version vs versionless.

| **Deployment type** | The way Elastic is deployed:<br><br>- **Serverless** (sometimes "Elastic Cloud Serverless")<br>- **Elastic Cloud Hosted**<br>- **Elastic Cloud on Kubernetes**<br>- **Elastic Cloud Enterprise**<br>- **Self-managed**<br><br>All deployment types other than **Serverless** are used to run a **{{stack-short}} <version>** flavor of Elasticsearch / Kibana. ECK and ECE also have their own versioning. For example, one can run a v9.0.0 deployment on ECE 4.0.
| **Project type** | The Serverless project types where a feature can be used - either **Elasticsearch**, **Security**, or **Observability** |
| **Other versioning schemes** | Elastic products or tools with a versioned component, where stack versioning is not followed.<br><br>E.g. clients, Elastic Common Schema |

:::{warning}
The self-managed deployment type and support for products or platforms other than {{stack-short}} does not exist yet. Contribute to the discussion in [#452](https://github.com/elastic/docs-builder/discussions/452)
:::

% TODO: Final term for "Stack"
% TODO: Final term for "Self-managed"

**How many facets do I need to use?**

> The role of these labels is providing a trust signal that the reader is viewing content that’s applicable to them. This means that the relevant labels should appear on all pages. However, we can choose to expose only one versioning plane on pages where only one plane is relevant:

Author: meant 'facet'

> The role of these labels is providing a trust signal that the reader is viewing content that’s applicable to them. This means that the relevant labels should appear on all pages. However, we can choose to expose only one versioning plane on pages where only one plane is relevant:

Author: I think this might be an edge-y enough case that it's better to just ask for people to always use labels. titles can always change as well - having the label metadata there will avoid information gaps

> The role of these labels is providing a trust signal that the reader is viewing content that’s applicable to them. This means that the relevant labels should appear on all pages. However, we can choose to expose only one versioning facet on pages where only one facet is relevant:

Author: see https://github.com/elastic/docs-builder/pull/451/files/7f71ed6cb11d9cb537a7d065af44d74f50087ff0#r1951791158

> The role of these labels is providing a trust signal that the reader is viewing content that’s applicable to them. This means that the relevant labels should appear on all pages. However, we can choose to expose only one versioning plane on pages where only one plane is relevant:

Author: ```suggestion
The role of these labels is providing a trust signal that the reader is viewing content that’s applicable to them. This means that the relevant labels should appear on all pages. However, we can choose to expose only one versioning facet on pages where only one facet is relevant:
```

* Depending on what you're documenting, you might need to include information from multiple facets. For example, when relevant features exist at both the stack and the deployment level, both of these groups might be used together (e.g. security, user management, and trust features differ both at the deployment level and at the stack version level).

> * In some sections, such as **Explore and analyze**, features generally only differ by stack flavor. In these cases, you can choose to include only this facet on the page.

Author: see https://github.com/elastic/docs-builder/pull/451/files/70db05b57d06e19e5286b0f893e2541152340611#r1950215255

### Defaults and hierarchy 

Do not assume a default deployment type, stack flavor, product version, or project type.

Treat all flavors and deployment types equally. Don't treat one as the "base" and the other as the "exception".

However, we should all display our flavors and versions in the same order for consistency. This order reflects organizational priorities.

**Flavors:**

1. Serverless
2. Stack <version>

**Deployment types:**

> 1. Serverless

Author: agree that it shouldn't appear twice. I expect the actual opportunities for it to appear twice are very low. 

depending on the design of the "applies to" section, it might need to appear twice when applicable (i.e. if we group labels by facet). otherwise we can just advise users to prefer the deployment type label when both facets apply.

2. Elastic Cloud Hosted
3. Elastic Cloud on Kubernetes
4. Elastic Cloud Enterprise
5. Self-managed

> When it comes to hierarchy of versions, always put the newest version first.

Author: represents our freshest features and most up-to-date processes. mimics our "newest at the top" system in dropdown menus.

### Versions and lifecycle states

In the V3 docs system, we currently document only our latest major versions (Stack 9.0+, ECE 4.0+, ECK 3.0+).

> Add version labels only when a feature was added as part of the major version release, or added in subsequent dot releases. Do not add version information to features added before these releases. For example, features added in 8.16 don't need an 8.16 label or a 9.0 label, but features added in 9.0 need a 9.0 label.

Reviewer 2: Why would we add an 9.0.0 label to an 8.16.0 feature?

> Add version labels only when a feature was added as part of the major version release, or added in subsequent dot releases. Do not add version information to features added before these releases. For example, features added in 8.16 don't need an 8.16 label or a 9.0 label, but features added in 9.0 need a 9.0 label.

Author: there was a question about this because the docs are 9.0.0+.

> From the latest major onward, the entire feature lifecycle should be represented. This means that you need to add separate labels for the following states:

Author: we can't say 9.0.0 here because these guidelines apply to all of our versioning systems.

* beta or technical preview

> * GA

Author: `ga` is used [under the hood](https://elastic.github.io/docs-builder/syntax/applies/)

> From the latest major onward, the entire feature lifecycle should be represented. This means that you need to add separate labels for the following states:

Author: ```suggestion
From the latest major version release onward, the entire feature lifecycle should be represented. This means that you need to add separate labels for the following states:
```

* deprecated
* removed

We hope to simplify or consolidate these lifecycle labels in future.

::::{warning}

> This feature doesn't currently work - currently, only one label will appear.

Author: currently, you can't have a separate label for `beta`, `ga`, and so on. not sure which one wins, but only one appears per deployment type (e.g. only one for "elastic stack" regardless of how many states you specify

::::

> ## Content patterns

Author: this is the parent page. it introduces the child.

This approach improves our documentation in several ways: 

* When a user arrives in our documentation from an outside source, they'll be likelier to land on a page or section that applies to them.
* There is a single "source of truth" for each feature, which helps us to maintain consistency, accuracy, and maintainability of our documentation over time, and avoids "drift" between multiple similar sets of documentation.
* Comparing and contrasting differences helps users to understand what is available to them, and improves awareness of the ways certain offerings might improve their experience.

::::{note}
This page documents the first version of our approach to documenting differences in versions and deployment types, using our current technologies. 
Our approach might change as additional documentation features become available.
::::

## Basic concepts and principles

* [Versioning facets](#versioning-facets)
* [Defaults and hierarchy](#defaults-and-hierarchy)
* [Versions and lifecycle states](#versions-and-lifecycle-states)

### Versioning facets
There are multiple facets to versioning that we need to consider: 

> * **Elasticsearch / Kibana flavor:** The feature base used for basic functionality. Either **Serverless** (sometimes "Elastic Cloud Serverless") or **Stack <version>**

Reviewer: Rather than limiting this to Elasticsearch and Kibana (since other things vary by deployment type too), can we generalize to something like this?
```suggestion
* **Platform or product versions:** For example, Elastic Stack 9.0.0. This is facet is currently not applicable in the context of Elastic Cloud Serverless.
```

% TODO: Final term for "Stack"
* **Deployment type:** The way Elastic is deployed: 
  * **Serverless** (sometimes "Elastic Cloud Serverless")
  * **Elastic Cloud Hosted**
  * **Elastic Cloud on Kubernetes**
  * **Elastic Cloud Enterprise**

>   * **Self-managed**

Reviewer: ```suggestion
  * **Self-managed**

:::{warning}
The self-managed deployment type and support for product/platforms other than Elastic Stack does not exist yet. Contribute to the discussion in [#452](https://github.com/elastic/docs-builder/discussions/452)
:::

```

>   All deployment types other than **Serverless** use the **Stack <version>** flavor of Elasticsearch / Kibana.

Reviewer: If my earlier suggestion to remove the "flavor" phrasing is accepted, this should be removed:
```suggestion
  All deployment types other than **Serverless** use the **Stack <version>** flavor of Elasticsearch / Kibana.
```

>   All deployment types other than **Serverless** use the **Stack <version>** flavor of Elasticsearch / Kibana.

Author: ^^^ good callout florent - ty

% TODO: Final term for "Self-managed"

* **Project type:** The Serverless project types where a feature can be used - either **Elasticsearch**, **Security**, or **Observability**.

> * **Other versioning schemes:** Elastic products or tools with a versioned component, where stack versioning is not followed. 

Reviewer: If my earlier suggestion about a more general "product/platform versions" phrasing is accepted, this is no longer required.
```suggestion
* **Other versioning schemes:** Elastic products or tools with a versioned component, where stack versioning is not followed. 
```

  
  E.g. clients, Elastic Common Schema

**How many facets do I need to use?**

The role of these labels is providing a trust signal that the reader is viewing content that’s applicable to them. This means that the relevant labels should appear on all pages. However, we can choose to expose only one versioning plane on pages where only one plane is relevant:

> * Depending on what you're documenting, you might need to include information from multiple facets. For example, when relevant features exist at both the stack and the deployment level, both of these groups might be used together (e.g. security, user management, trust).

Reviewer: ```suggestion
* Depending on what you're documenting, you might need to include information from multiple facets. For example, when a page pertains to a specific deployment type and product version ( for example, Cloud deployments for a particular Elastic Stack version number), specify all of the relevant facets.
```

> * In some sections, such as **Explore and analyze**, features generally only differ by Elasticsearch flavor. In these cases, you can choose to include only this facet on the page.

Reviewer: > * In some sections, such as **Explore and analyze**, features generally only differ by Elasticsearch flavor. In these cases, you can choose to include only this facet on the page.

This implies that any page that doesn't specify one of the facets applies to all the rest. If we're going to go this very clear.

Author: not 100% sure how to address this

### Defaults and hierarchy 

> You should not assume a default deployment type or Elasticsearch / Kibana flavor.

Reviewer: I think the previous section related to the frontmatter and this one relates to the content style. This is my attempt to tie them together.

```suggestion
Do not assume that there is a default deployment type or product version.
Rather, write every page to apply to every deployment type and product version. If that broad applicability is not true, make it clear in the frontmatter.
```

However, we should all display our flavors and versions in the same order for consistency. This order reflects organizational priorities.

**Flavors:**

1. Serverless
2. Stack

**Deployment types:**

1. Serverless
2. Elastic Cloud Hosted
3. Elastic Cloud on Kubernetes
4. Elastic Cloud Enterprise
5. Self-managed

When it comes to hierarchy of versions, always put the newest version first.

### Versions and lifecycle states

In the V3 docs system, we currently document only our latest major versions (Stack 9.0+, ECE 4.0+, ECK 3.0+).

Add version labels only when a feature was added as part of the major version release, or added in subsequent dot releases. Do not add version information to features added before these releases. For example, features added in 8.16 don't need an 8.16 label or a 9.0 label, but features added in 9.0 need a 9.0 label.

From the latest major onward, the entire feature lifecycle should be represented. This means that you need to add separate labels for the following states:

* beta or technical preview
* GA
* deprecated
* removed

We hope to simplify or consolidate these lifecycle labels in future.

::::{warning}

> This feature doesn't currently work - currently, only one label will appear.

Reviewer: > This feature doesn't currently work - currently, only one label will appear.

I think we should put a warning like this about the lack of an "on-prem / self-managed" deployment type too.
